nemo/collections/tts/modules/audio_codec_modules.py

```diff
@@ -167,6 +174,318 @@ def forward(self, audio_real, audio_gen):
+def zero_mean_unit_var_norm(input_values):
+    """
+    Normalized to have zero mean and unit variance
+    """
+    normed_input_values = (input_values - input_values.mean(dim=1).unsqueeze(-1)) / torch.sqrt(
+        input_values.var(dim=1).unsqueeze(-1) + 1e-7
+    )
+    return normed_input_values
+
+
+class PhonemeASR(NeuralModule):
+    def __init__(self, input_sr=22050, model_sr=16000):
+        super().__init__()
+        # self.processor = Wav2Vec2Processor.from_pretrained("facebook/wav2vec2-lv-60-espeak-cv-ft") # processor are not grad friendly
+        self.model = Wav2Vec2ForCTC.from_pretrained("facebook/wav2vec2-lv-60-espeak-cv-ft")
+        if HAVE_TORCHAUDIO:
+            self.resample = torchaudio.transforms.Resample(input_sr, model_sr)
+        else:
+            self.resample = None
+
+    def forward(self, audio):
+        audio = self.resample(audio)
+        # input_values = self.processor(audio, return_tensors="pt").input_values
+        input_values = zero_mean_unit_var_norm(audio)
+        logits = self.model(input_values).logits
+        predicted_ids = torch.argmax(logits, dim=-1)
+
+        # transcription = self.processor.batch_decode(predicted_ids)
+        # print("GT Phonemes:", transcription[0])
+        # print("GEN Phonemes:", transcription[len(transcription)//2])
+        return logits, predicted_ids
+
+
+##############
+# Speaker encoder #
+##############
+def load_fsspec(path: str, map_location: str = None, **kwargs):
+    """Like torch.load but can load from other locations (e.g. s3:// , gs://).
+
+    Args:
+        path: Any path or url supported by fsspec.
+        map_location: torch.device or str.
+        cache: If True, cache a remote file locally for subsequent calls. It is cached under `get_user_data_dir()/tts_cache`. Defaults to True.
+        **kwargs: Keyword arguments forwarded to torch.load.
+
+    Returns:
+        Object stored in path.
+    """
+    is_local = os.path.isdir(path) or os.path.isfile(path)
+    if is_local:
+        return torch.load(path, map_location=map_location, **kwargs)
+    else:
+        if HAVE_FSSPEC:
+            with fsspec.open(path, "rb") as f:
+                return torch.load(f, map_location=map_location, **kwargs)
+        else:
+            logging.error('Could not import fsspec. Loading a checkpoint link is not supported!')
+            raise ModuleNotFoundError(f"fsspec is not installed but is necessary to download remote checkpoints !!")
+
+
+class PreEmphasis(NeuralModule):
+    def __init__(self, coefficient=0.97):
+        super().__init__()
+        self.coefficient = coefficient
+        self.register_buffer("filter", torch.FloatTensor([-self.coefficient, 1.0]).unsqueeze(0).unsqueeze(0))
+
+    def forward(self, x):
+        assert len(x.size()) == 2
+
+        x = torch.nn.functional.pad(x.unsqueeze(1), (1, 0), "reflect")
+        return torch.nn.functional.conv1d(x, self.filter).squeeze(1)
```

Comment:
What's the difference between this and https://github.com/NVIDIA/NeMo/blob/e5899c37a27952bf003c2b8d082ebee68fbb4f7f/nemo/collections/asr/parts/preprocessing/features.py#L433

Reply:
They are similar, but the current implementation in NeMo preserves the first value of each vector, whereas the implementation in this class modifies the entire input. Since the Speaker Encoder model was trained using this new implementation, I don’t think we can replace it.